### DISCUSSION TITLE
Pipelining TMA async memcpy

### DISCUSSION THREAD

**Discussion by rickyyx** [NVIDIA/cccl#1789] [libcu++]
Hey,  I was trying to speed up my kernel that invovles memcpy from global + some computing, looking at the official documentation of async memcpy, I see there are multiple ways of doing async memcpy:

pipelining with memcpy_async [doc](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#asynchronous-data-copies-using-cuda-pipeline)
using the TMA related memcpy_async [doc](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#asynchronous-data-copies-using-tensor-memory-access-tma)

The pipelined async_memcpy seems to be natively support with memcpy_async from cuda/pipeline taking in a pipeline instance. However, there seems to be no similar API for the TMA’s group of APIs.

Is one able to speed up the TMA async mecmpy with pipeline with current interfaces and APIs?

Thanks

**Comment by ahendriksen** [comment]
Hi @rickyyx, 

Good question! 

> The pipelined async_memcpy seems to be natively support with memcpy_async from cuda/pipeline taking in a pipeline instance. However, there seems to be no similar API for the TMA’s group of APIs.

This is correct.

> Is one able to speed up the TMA async mecmpy with pipeline with current interfaces and APIs?

Unfortunately, not (yet). 

What is your use case? Does the pipeline example using mempcy async from [the CUDA blog](https://developer.nvidia.com/blog/controlling-data-movement-to-boost-performance-on-ampere-architecture/#overlapping_global-to-shared_copies_with_compute) speed up your kernel?
